### PR TITLE
libc++ update

### PR DESCRIPTION
**Pull request by cynecx**
See https://groups.google.com/forum/#!topic/emscripten-discuss/g7Xqry-1OBU for details.

This change includes the whole libc++ library update to SVN revision 268153 (upstream) plus some fixes so emscripten will be able to compile the updated libc++.

I have done some local tests (mainly personal projects), nothing heavy so that testing is still required.
It would be great if anyone else could run the whole test suite on this though.

### CODE REVIEW

**Comment by kripken** [comment]
Nice!

I'm running the test suite now.

**Comment by kripken** [comment]
I think we still need to separate `C_INCLUDE_PATHS` from c++, as when building C code, we shouldn't have libc++ in the path.

**Comment by cynecx** [comment]
@kripken I undid the change and reordered later.
I have a questions though. C_INCLUDE_PATHS and CXX_INCLUDE_PATHS only reference is this line:

    EMSDK_OPTS = C_OPTS + include_directive(CXX_INCLUDE_PATHS) + include_directive(C_INCLUDE_PATHS)

So does it matter whether we are building C code or C++ because it will have libc++ headers included?
Am I missing here something?

**Comment by kripken** [comment]
I see two failures, on `test_emulate_function_pointer_casts, test_iostream_ctors`.

**Comment by kripken** [comment]
@cynecx: they might also be used in a program importing `shared.py`, like `emcc.py` or something in `tools/`. If there are no uses there, then it would be safe.

**Comment by cynecx** [comment]
@kripken I am currently investigating this. By my first look it seems we have issues again with headers.
In the `test_emulate_function_pointer_casts` case, we are including math.h.

But we have two math headers one in libc and one in libcxx.
Libc (math.h) defines acos as:

    double      acos(double);

On the other hand libcxx:

    inline float       acos(float __lcpp_x) noexcept       {return acosf(__lcpp_x);}
    inline long double acos(long double __lcpp_x) noexcept {return acosl(__lcpp_x);}

    template <class _A1>
    inline typename std::enable_if<std::is_integral<_A1>::value, double>::type
    acos(_A1 __lcpp_x) noexcept {return acos((double)__lcpp_x);}

The difference here is that libcxx uses definitions declared inline so that why you can't do a function pointer conversion into a different function pointer.

So we are having trouble again with the headers.

Uhm. I think I will figure out something else, this seems very hacky to continue.

**Comment by cynecx** [comment]
A small update. Actually it seems that it's not our fault that these tests aren't working anymore, I think it's upstream. As I explained earlier libcxx is defining inline overloads of acos and other functions which prevents that function pointer conversion.

So I would say that this patch is completely fine but we have upstream issues. Although I wouldn't call this upstream issue because the compile error (`address of overloaded function 'acos' does not match required type`) is actually correct.

What do you think @kripken?

**Comment by cynecx** [comment]
I fixed the test by using a newly defined function which acts as a proxy between acos and main.
The semantics of the test should remain the same though.

**Comment by kripken** [comment]
Honestly, that problem goes beyond my understanding of C++ :) I'm not sure what is best to do there. Perhaps @juj can help?

Is the problem related to the reordering of includes, btw? I.e. would it happen before with the order changed?

**Comment by cynecx** [comment]
@kripken Sorry for the confusion.

So the issue is that libcxx could only compile if we have the CXX include dirs before libc ones because libcxx is using `#include_next` hackery (libcxx is modifying libc headers by providing them in libcxx). This resulted into a properly working libcxx but then you caught a test failure which relies on an one declared extern function acos and since libcxx is kinda "overloading" the math.h header and shadowing the acos declaration with other overloads it's simply not possible to cast acos function pointer (because we now have an overload set) to a not suitable signature function pointer (`address of overloaded function 'acos' does not match required type`).

In my opinion it is the best to modify the test where the semantic stays, as I did in 466172a.

File: system/lib/libcxx/iostream.cpp [hunk 1]
```diff
@@ -41,7 +41,7 @@ static mbstate_t mb_wcerr;
 _ALIGNAS_TYPE (ostream)  _LIBCPP_FUNC_VIS char clog[sizeof(ostream)];
 _ALIGNAS_TYPE (wostream) _LIBCPP_FUNC_VIS char wclog[sizeof(wostream)];
 
-ios_base::Init __start_std_streams;
+ios_base::Init __attribute__((init_priority(101))) __start_std_streams;
```
Comment (kripken) [inline system/lib/libcxx/iostream.cpp]: there was also a comment in the commit from #3824, we should keep that as well.

**Comment by kripken** [comment]
Ok, I see about the `acos` issue. Yes, I think fixing the test is the right thing, as you did. The issue is indeed the overloading. Please add a comment with that to the test, as it looks very odd otherwise :)

**Comment by kripken** [comment]
Please also update `system/include/libcxx/readme.txt`.

**Comment by cynecx** [comment]
@kripken What should be updated in `system/include/libcxx/readme.txt`?.

**Comment by kripken** [comment]
That file says which revision of libc++ it is, and what date, so this update to those files should say the new revision and date.

**Comment by cynecx** [comment]
Actually, it should be already updated. Can you re-check the file?


On 08 May 2016, at 04:46, Alon Zakai <notifications@github.com<mailto:notifications@github.com>> wrote:


That file says which revision of libc++ it is, and what date, so this update to those files should say the new revision and date.

-

**Comment by cynecx** [comment]
https://github.com/cynecx/emscripten/blob/libcxx_update/system/lib/libcxx/readme.txt

**Comment by kripken** [comment]
Sorry, yeah, I see now it has already been updated. It looks like the github UI doesn't show all the files, even when searching in the list, strange.

Another concern I had is that you moved `new.cpp` from libcxxabi to libcxx. I think the reason we had it as it was before is that it was an optimization - a program just doing `new/delete` but not using other libc++ stuff would not need to link in libc++, which is large and takes  time. I measured this, -O0 on a trivial program doing just `new`, it is 18% slower, on -O3 it's 10%. On BananaBread, a more realistic-sized project, it's 2%. Anyhow your change is probably for the best, even if it makes some small trivial programs a little slower. Perhaps we can optimize this in a cleaner way later, for now this seems fine.

**Comment by kripken** [comment]
Thanks for doing this! :)

**Comment by cynecx** [comment]
@kripken No problem :).

> Another concern I had is that you moved new.cpp from libcxxabi to libcxx. [...]

Actually we could add pre-processor hackery to new.cpp (`#ifdef LIBCXX_BUILDING_LIBCXXABI ...`) so only weak definitions of libcxxabi are available. Although I am not sure if this is clean xD.

The only advantage I see here is that pulling newer sources from upstream would be easier because we could then simply replace files without actually dealing with merging conflicts (and of course re-adding patches).

But yeah I agree, this should be addressed later :).

**Comment by qis** [comment]
@cynecx Hi, I would like to replicate your work and upgrade libcxx to LLVM 4.0.0 with `-std=c++1z` support. Can you give me a hint on where to start? Any notes/documentation you can provide or link to? Is this a feasible task to do for a single person in under 100 hours?

@kripken If you have more information, I will welcome it as well.

**Comment by dschuff** [comment]
@qis I think so; in fact I would expect it to be easier than upgrading LLVM itself. The sources are in `system/lib/libcxx` and it fortunately doesn't look like we have too many differences from upstream. Probably most of the work would involve sussing out issues with the way it's built, and maybe tests that have some dependence on something weird. Also @jgravelle-google recently updated libc... where there any changes to how symbols are updated?

**Comment by jgravelle-google** [comment]
My understanding is that our libc is built as a bitcode file, and its symbols are read from that directly, so I didn't need to do anything to symbol files. libcxx I believe uses archive files, though I don't have any real insight as to how.

**Comment by qis** [comment]
I'm pretty sure that symbol files must be changed since all my attempts to use `em++ -std=c++1z` and binaryen failed with

```
Import #0 module="env" function="_ZSt18uncaught_exceptionv" error:
  function import requires a callable
```

Which shouldn't exist in the first place since `std::uncaught_exception` was deprecated in favor of `std::uncaught_exceptions`.

**Comment by dschuff** [comment]
Yeah the symbol files will have to be updated. @jgravelle-google meant that they *didnt* have to be update for libc because libc is just a giant bitcode file instead of an archive like the rest of the libraries. The format of the symbol files looks like just the output of `llvm-nm` run over the library archive.

**Comment by qis** [comment]
@dschuff This is helpful, thank you. I'll start tomorrow.

**Pull request merged**